### PR TITLE
Add Incremental Indexing v1

### PR DESCRIPTION
**Pull request by AlonsoGuevara**
## Description

Add Incremental Indexing support.

## Usage
Drop new files into the already existing input folder and update settings.yaml

In settings.yaml

```yaml
storage:
  type: file
  base_dir: "path/to/existing/artifacts"

update_index_storage:
  type: file
  base_dir: "path/to/create/new/index"
  
```

## Related Issues

#741 

## Checklist

- [x] I have tested these changes locally.
- [x] I have reviewed the code changes.
- [ ] I have updated the documentation (if necessary).
- [ ] I have added appropriate unit tests (if applicable).

## Additional Notes

[Add any additional notes or context that may be helpful for the reviewer(s).]

### CODE REVIEW

File: graphrag/config/create_graphrag_config.py [hunk 1]
```diff
@@ -375,6 +375,25 @@ def hydrate_parallelization_params(
...
+                    connection_string=reader.str(Fragment.conn_string),
+                    storage_account_blob_url=reader.str(
+                        Fragment.storage_account_blob_url
+                    ),
+                    container_name=reader.str(Fragment.container_name),
+                    base_dir=reader.str(Fragment.base_dir)
+                    or defs.UPDATE_STORAGE_BASE_DIR,
+                )
+            else:
+                update_index_storage_model = None
```
Comment (andresmor-ms) [inline graphrag/config/create_graphrag_config.py:396]: you could remove this else by moving the `update_index_storage_model = None` above the if :), could even go outside the with statement.

File: graphrag/index/flows/create_final_communities.py [hunk 2]
```diff
@@ -61,6 +63,12 @@ def create_final_communities(
 
     filtered["title"] = "Community " + filtered["id"].astype(str)
 
+    # Add period timestamp to the community reports
+    filtered["period"] = datetime.now(timezone.utc).date().isoformat()
+
+    # Add size of the community
+    filtered["size"] = filtered.loc[:, "text_unit_ids"].apply(lambda x: len(x))
```
Comment (andresmor-ms) [inline graphrag/index/flows/create_final_communities.py:70]: I think you can change it to `.apply(len)` no need for the lambda

Comment (AlonsoGuevara) [reply graphrag/index/flows/create_final_communities.py:70]: I'll submit another PR optimizing all this calculations. This change was to attack some warnings being output

File: graphrag/config/create_graphrag_config.py (hunk 1 above)
Comment (AlonsoGuevara) [reply graphrag/config/create_graphrag_config.py:396]: I had it like that but the issue was that the interpreter was assuming this to always be set. So it greyed out some of my none checks and was driving me nuts.
I can try again by setting explicit typings.

**Review by andresmor-ms — DISMISSED** [review]

**Review by andresmor-ms — APPROVED** [review]

**Comment by shamalgithub** [comment]
Thank you for this feature update ! 

Based on the instructions , i belive all one has to do is to add a new file to the 'input' folder and update the settings.yaml file with the new output file name.  
image

Looking at the issue , the proposed fix suggest that a graphrag.append command would be added to work with increamental indexing but that doesnot seem to  be present. 
image

I might be missing something here. Can someone kindly point out the correct way to do incremental indexing ?

**Comment by joycerlz** [comment]
I also have the same question about the graphrag.append command, is it a valid command yet? Or we just add the file and update settings.yaml, no need to run additional commands? Any response/discussion would be appreciated, thanks!

**Pull request merged**